### PR TITLE
Record stack trace on service test failures.

### PR DESCRIPTION
**Pull request by robertzk**
PIVOTAL: https://www.pivotaltracker.com/story/show/112374031

### CODE REVIEW

**Comment by robertzk** [comment]
@khalidharun This is inspired by [evaluate](https://github.com/hadley/evaluate)'s [stack trace capture](https://github.com/hadley/evaluate/blob/eb304135a5370a1ffad80717c50fe152ebeb7b4f/R/traceback.r).

**Comment by khalidharun** [comment]
@robertzk A little real-life test here gives me a stack trace but nothing really useful:
```
> with_mock(`avant:::get_db_conn` = function(...) stop("force error here"), run_servicetests("run_query_us on loans"))
> test_failures[[1]]$actual$calls
[[1]]
with_mock(`avant:::get_db_conn` = function(...) stop("force error here"),
    run_servicetests("run_query_us on loans"))

[[2]]
eval(expression, parent.frame())

[[3]]
eval(expr, envir, enclos)

[[4]]
run_servicetests("run_query_us on loans")

[[5]]
servicetests:::test_service_calls(keys = keys, return.type = return.type,
    on.pass_all = pass_all_fn(test_name, is_cronjob), on.fail_any = fail_any_fn(test_name,
        is_cronjob), on.fail_each = fail_each_fn, on.pass_each = pass_each_fn)

[[6]]
lapply(reg, test_service, envir = parent.frame(), ...)

[[7]]
FUN(X[[1L]], ...)

[[8]]
tryCatch(withCallingHandlers({
    res <- do.call(eval(parse(text = obj$fname), envir = envir),
        obj$params, envir = envir)
    match_fn <- obj$match_fn %||% "all.equal"
    passed <- isTRUE(do.call(match_fn, list(res, obj$expected)))
    list(passed = passed, actual = res)
}, error = capture_calls), error = function(e) {
    list(passed = FALSE, actual = e)
})

[[9]]
tryCatchList(expr, classes, parentenv, handlers)

[[10]]
tryCatchOne(expr, names, parentenv, handlers[[1L]])

[[11]]
doTryCatch(return(expr), name, parentenv, handler)
```

**Comment by khalidharun** [comment]
@robertzk Removing `head(...)` around `sys.calls()` gives us the full trace!

```
> test_failures[[1]]$actual$calls
[[1]]
with_mock(`avant:::get_db_conn` = function(...) stop("force error here"),
    run_servicetests("run_query_us on loans"))

[[2]]
eval(expression, parent.frame())

[[3]]
eval(expr, envir, enclos)

[[4]]
run_servicetests("run_query_us on loans")

[[5]]
servicetests:::test_service_calls(keys = keys, return.type = return.type,
    on.pass_all = pass_all_fn(test_name, is_cronjob), on.fail_any = fail_any_fn(test_name,
        is_cronjob), on.fail_each = fail_each_fn, on.pass_each = pass_each_fn)

[[6]]
lapply(reg, test_service, envir = parent.frame(), ...)

[[7]]
FUN(X[[1L]], ...)

[[8]]
tryCatch(withCallingHandlers({
    res <- do.call(eval(parse(text = obj$fname), envir = envir),
        obj$params, envir = envir)
    match_fn <- obj$match_fn %||% "all.equal"
    passed <- isTRUE(do.call(match_fn, list(res, obj$expected)))
    list(passed = passed, actual = res)
}, error = capture_calls), error = function(e) {
    list(passed = FALSE, actual = e)
})

[[9]]
tryCatchList(expr, classes, parentenv, handlers)

[[10]]
tryCatchOne(expr, names, parentenv, handlers[[1L]])

[[11]]
doTryCatch(return(expr), name, parentenv, handler)

[[12]]
withCallingHandlers({
    res <- do.call(eval(parse(text = obj$fname), envir = envir),
        obj$params, envir = envir)
    match_fn <- obj$match_fn %||% "all.equal"
    passed <- isTRUE(do.call(match_fn, list(res, obj$expected)))
    list(passed = passed, actual = res)
}, error = capture_calls)

[[13]]
do.call(eval(parse(text = obj$fname), envir = envir), obj$params,
    envir = envir)

[[14]]
(function (query)
{
    run_query_in_locale(query, locale = "US")
})("select id from loans where id in (544983, 317744, 508419, 672385, 1356229)")

[[15]]
run_query_in_locale(query, locale = "US")

[[16]]
DBI::dbGetQuery(get_db_conn(locale), query)

[[17]]
.handleSimpleError(function (e)
{
    e$calls <- sys.calls()
    signalCondition(e)
}, "error in evaluating the argument 'conn' in selecting a method for function 'dbGetQuery': ",
    quote(DBI::dbGetQuery(get_db_conn(locale), query)))

[[18]]
h(simpleError(msg, call))
```

**Comment by khalidharun** [comment]
@robertzk wanna get this merged in?

File: R/servicetests.R [hunk 1]
```diff
@@ -141,7 +141,7 @@ test_service <- function(obj, envir = parent.frame(), on.pass_each = NULL, on.fa
         passed = passed,
         actual = res
       )
-    },
+    }, error = capture_calls), # Capture stack trace on errors.
     error = function(e) {
```
Comment (khalidharun) [inline R/servicetests.R]: @robertzk this section looks really odd.  Do we need both `tryCatch` and `withCallingHandlers`?

Comment (peterhurford) [inline R/servicetests.R]: `tryCatch` is just a wrapper around `withCallingHandlers` anyway.

**Comment by peterhurford** [comment]
@robertzk @khalidharun Ping?

File: R/servicetests.R (hunk 1 above)
Comment (robertzk) [inline R/servicetests.R]: Good point.

Comment (robertzk) [inline R/servicetests.R]: Actually, yes, we do want both a `tryCatch` and `withCallingHandlers`: note that `capture_calls` signals

Comment (robertzk) [inline R/servicetests.R]:
```r
> foo <- function(a) {
+   bar(a)
+ }
> bar <- function(a) {
+   stop("omg")
+ }
> wtf <- function() {
+   tryCatch(withCallingHandlers(error = function(e) { e$calls <- sys.calls(); signalCondition(e) },
+     foo(1)), error = function(e) { list(error = e, caught = TRUE) })
+ }
> wtf()
$error
<simpleError in bar(a): omg>

$caught
[1] TRUE

> wtf()$error$calls
[[1]]
wtf()

[[2]]
tryCatch(withCallingHandlers(error = function(e) {
    e$calls <- sys.calls()
    signalCondition(e)
}, foo(1)), error = function(e) {
    list(error = e, caught = TRUE)
})
# etc...
```

**Pull request merged**